habitat/core/env.py

```diff
@@ -423,3 +424,232 @@ def __enter__(self):
+        """
+        if (
+            self._task_cycling_behavior
+            == CHANGE_TASK_CYCLE_BEHAVIOUR.ORDER.name
+        ):
+            self._curr_task_idx = (self._curr_task_idx + 1) % len(self._tasks)
+        elif (
+            self._task_cycling_behavior
+            == CHANGE_TASK_CYCLE_BEHAVIOUR.RANDOM.name
+        ):
+            # sample from other tasks with equal probability
+            self._curr_task_idx = np.random.choice(
+                [
+                    i
+                    for i in range(len(self._tasks))
+                    if i != self._curr_task_idx
+                ],
+                1,
+            ).item()
+
+        self._task = self._tasks[self._curr_task_idx]
+        # update episode iterator with task dataset
+        iter_option_dict = {
+            k.lower(): v
+            for k, v in self._config.ENVIRONMENT.ITERATOR_OPTIONS.items()
+        }
+        iter_option_dict["seed"] = self._config.SEED
+        self._episode_iterator: EpisodeIterator = (
+            self._task._dataset.get_episode_iterator(**iter_option_dict)
+        )
+
+        self.action_space = self._task.action_space
+        self._task_label = self._task._config.get(
+            "TASK_LABEL", self._curr_task_idx
+        )
+        # handle mid-episode change of task
+        if not is_reset:
+            # keep current position and sim state only if scene does not change
+            # if same exact task is passed, you'll get same episode back
+            if (
+                self.current_episode.scene_id
+                == self._episode_iterator.episodes[0].scene_id
+            ):
+                self._episode_iterator.episodes[
+                    0
+                ].start_position = self.current_episode.start_position
+                self._episode_iterator.episodes[
+                    0
+                ].start_rotation = self.current_episode.start_rotation
+                # self._episode_iterator._iterator = iter(self._episode_iterator.episodes)
+                # self.reconfigure(self._config)
+
+                # observations = self.task.reset(episode=self.current_episode)
+                # update current episode
+                self._current_episode = next(self._episode_iterator)
+                # reset prev position
+                self._task.measurements.reset_measures(
+                    episode=self.current_episode, task=self.task
+                )
+            else:
+                # scene is different, can't keep old position, end episode
+                self._episode_over = True
+
+    def step(
+        self, action: Union[int, str, Dict[str, Any]], **kwargs
+    ) -> Observations:
+        obs = super().step(action, **kwargs)
+        # check task change behaviour here
+        self._cumulative_steps_counter += 1
+        if self._check_change_task():
+            # task needs to be changed
+            self._change_task()
+
+        obs["task_label"] = self._task_label
```

Comment:
The observation space needs to be updated to expect this `task_label` observation.

Reply:
Good point thanks, I'll fix that.

Reply:
I guess I'll pass the current task index instead of `task_label` here with a `Discrete` space, I don't think motivations for a task label are as strong as to define a custom gym space with support to arbitrary strings here, due to warnings in https://github.com/openai/gym/blob/master/gym/spaces/space.py.